### ISSUE TITLE
How to set user context on request

### ISSUE THREAD

**Issue by kendallroth** [getsentry/sentry-python#161]
I've been trying to get user context working for the past few days with no success unfortunately. I have created a Django middleware that does run with each received request and correctly gets the user from the request context, but the scope doesn't get updated properly. No matter what I try, the user data is never sent back to Sentry.io.

**See the bottom code example**

### Issue with `send_default_pii`

Additionally, I noticed that setting `send_default_pii=True` appears to actually send less tags to Sentry. For example, with `send_default_pii=False` I get these tags:

![image](https://user-images.githubusercontent.com/12562420/47806466-3b5d8500-dd10-11e8-9288-2f4ee3e42b4d.png)

However, with `send_default_pii=True` I only get these tags, which is backwards from what I would expect:

![image](https://user-images.githubusercontent.com/12562420/47806684-b45cdc80-dd10-11e8-8bb6-54ce4d9e19f7.png)

The docs mention the following:
> Personally identifiable information (such as user ids, usernames, cookies, authorization headers, ip addresses) is excluded unless send_default_pii is set to True.

Also from the docs (same page):
> If you use django.contrib.auth and have set send_default_pii=True in your call to init, user data (current user id, email address, username) is attached to the event

### Details

- **Server**: Django with Graphene

```python
# settings.py

import sentry_sdk
from sentry_sdk.integrations.django import DjangoIntegration

MIDDLWARE = [
    ..
   'django.contrib.auth.middleware.AuthenticationMiddleware',
    # Must be in both Django middleware and Graphene middleware (behaves as both)
    #   Needs to be in middleware before sentry middleware to set user properly before trying to set Sentry context
    'graphql_jwt.middleware.JSONWebTokenMiddleware',
    ...
    # Middleware to intercept requests and add Sentry user to context
    'util.middleware.SentryMiddleware'
]

GRAPHENE = {
    'SCHEMA': '....',
    'MIDDLEWARE': [
        ...
        # Must be in both Django middleware and Graphene middleware (behaves as both)
        #   Apparently meant to be Graphene middleware (as according to their docs)?
        'graphql_jwt.middleware.JSONWebTokenMiddleware',
    ]
}

...

# Sentry error logging (at end of file)
sentry_sdk.init(
    dsn=os.getenv("API_SENTRY_DSN"),
    environment=os.getenv("ENV_NAME"),
    integrations=[DjangoIntegration()],
    #  QUESTION: Setting to True appears to not send as much information to Sentry (tags)?
    send_default_pii=True
)
```

Am I initializing the SDK in the right place (docs mention the `settings.py` file)?

```python
# middleware.py

from sentry_sdk import configure_scope

class SentryMiddleware:
  def __init__(self, get_response):
    """Configuration and initialization."""
    self.get_response = get_response

  def __call__(self, request):
    """Hook in each request before view is resolved."""
    # NOTE: User does exist here (if authenticated) and is set properly on the request
    if request.user and request.user.is_authenticated:
      user = {
        "email": request.user.email,
        "id": request.user.id,
        "name": request.user.profile.full_name,
        "username": request.user.username
      }

      # Does this actually modify the Sentry scope? It doesn't appear to send user details to Sentry...
      with configure_scope() as scope:
        scope.user = request.user

    response = self.get_response(request)

    return response
```

Am I using the middleware correctly? Any help would be appreciated

**Comment by untitaker** [comment]
Hi @kendallroth, you are setting `scope.user` to `request.user` instead of using the dictionary you assembed a few lines before. I would fix that first.

Regardless of that the DjangoIntegration should already do roughly the same thing you do in the middleware, automatically.

I can't reproduce the issue with the user-agent tag. That tag is set by sentry itself when we send request headers as part of the event, and we do not touch that header at all regardless of `send_default_pii`. You can see here what we do with headers: https://github.com/getsentry/sentry-python/blob/3fe8712842d03a3dfea7fdd10719777f1ce56f78/sentry_sdk/integrations/_wsgi.py#L251-L259

Another question though, which server are you using to test this? I am unfortunately not very familiar with GraphQL and the ecosystem around it but I assume you would still use a regular WSGI server for all of this?

**Comment by kendallroth** [comment]
You are completely correct about `scope.user`, my apologies (didn't change that back when assembling example).

I figured that was what the `DjangoIntegration` would do, but since it hadn't worked I thought I'd try it myself to see what was going wrong. I was able to verify that `request.user` was definitely set, but couldn't attach to the Sentry scope successfully.

The only difference between the two images (showing tag diff) was `send_default_pii`, which strikes me as very strange. I can't understand what may have caused that then, as we only use default other middleware.

As far as server, I'm not entirely sure what server we are using TBH, but our Docker file contains this:

```bash
EXPOSE ####
CMD ["./manage.py", "runserver", "0.0.0.0:####"]
ENTRYPOINT ["pipenv", "run"]
```

Our `urls.py` file renders the Graphene "view":

```python
from django.urls import path
from django.conf.urls import url
from graphene_django.views import GraphQLView

urlpatterns = [
    path('admin/', admin.site.urls),
    url(r'^graphql', csrf_exempt(
        GraphQLView.as_view(graphiql=True, schema=schema)
    ))
]
```

Additionally, we do have a file `wsgi.py` in the same directory as `settings.py`:

```python
"""
WSGI config for [APP_NAME] project.
It exposes the WSGI callable as a module-level variable named ``application``.
For more information on this file, see
https://docs.djangoproject.com/en/2.0/howto/deployment/wsgi/
"""

import os
from django.core.wsgi import get_wsgi_application

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "[APP_NAME].settings")

application = get_wsgi_application()
```

**Comment by untitaker** [comment]
sorry for the slow response. While we have seen cases of missing request data, this behavior does not make sense to me at all, especially the correlation with `send_default_pii`. I'll investigate in the upcoming week. In the meantime (if you have enough time to spare) you could try to delete chunks of your application unless you have a sufficiently small app that still exhibits the problem. The biggest uncertainty for me here is whether GraphQL is required to reproduce this problem.

**Comment by kendallroth** [comment]
Ah, therein lies the problem. We're under a very strict deadline for the next 3 weeks and I won't be able to create a reproducible example in that time period unfortunately. As far as GraphQL is concerned, that's a very interesting point (could possibly be `graphene`) and will probably be the first thing I try when I do have time.

**Comment by untitaker** [comment]
@kendallroth if that's 0.2.0 of the SDK you're talking about I would recommend jumping directly to 0.5.0. There were no API changes but very critical bug fixes.

**Comment by kendallroth** [comment]
Sorry, I'm not sure why I said `0.2.0`... We are using version `0.5.0`, which is where the issue appeared to be fixed.

`sentry-sdk = ">=0.5,<1"`

**Comment by untitaker** [comment]
Ah yes then a fix probably slipped in there. We found a lot of bugs around that time
